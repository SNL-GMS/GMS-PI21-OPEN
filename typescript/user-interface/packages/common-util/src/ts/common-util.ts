import * as Timer from './common-util/performance-monitoring-util';

export * from './common-util/array-util';
export * from './common-util/compose-util';
export * from './common-util/delay-execution-util';
export * from './common-util/environment-util';
export * from './common-util/file-system-util';
export * from './common-util/formatting-util';
export * from './common-util/json-util';
export * from './common-util/logger';
export * from './common-util/min-max-util';
export * from './common-util/random-number-util';
export * from './common-util/range-util';
export * from './common-util/record-util';
export * from './common-util/sleep-util';
export * from './common-util/string-util';
export * from './common-util/time-util';
export * from './common-util/type-util';
export * from './common-util/uuid-util';
export * from './common-util/window-util';
export { KeyValue } from './types/key-value';
export { ValueType } from './types/value-type';
export { Timer };
