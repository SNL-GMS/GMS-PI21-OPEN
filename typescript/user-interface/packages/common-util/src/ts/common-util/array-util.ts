import filter from 'lodash/filter';
import includes from 'lodash/includes';
import uniq from 'lodash/uniq';

/**
 * @returns whether the two lists are unique
 */
export const isUnique = (a: unknown[] | undefined): boolean => a && uniq(a).length === a.length;

/** Gets a list the items that are duplicates */
export const getDuplicates = (arr: unknown[]): unknown[] =>
  filter(arr, (val, i: number, iteratee) => includes(iteratee, val, i + 1));
