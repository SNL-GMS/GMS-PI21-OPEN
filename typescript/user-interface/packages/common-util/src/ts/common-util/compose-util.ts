/**
 * Composes a function that invokes the given functions from right to left.
 *
 * @param funcs the functions to compose
 */
// TODO Replace this with lodash's flowRight function
// eslint-disable-next-line @typescript-eslint/ban-types, @typescript-eslint/explicit-module-boundary-types
export function compose(...funcs: Function[]) {
  funcs.reverse();
  return (...args: unknown[]) => {
    const [firstFunction, ...restFunctions] = funcs;
    // eslint-disable-next-line prefer-spread
    let result = firstFunction.apply(null, args);
    restFunctions.forEach(fnc => {
      result = fnc.call(null, result);
    });
    return result;
  };
}
