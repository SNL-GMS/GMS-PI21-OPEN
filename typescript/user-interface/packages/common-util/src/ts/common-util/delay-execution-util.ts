/**
 * Delay invoking a function for some number of milliseconds.
 *
 * @param func the function to call on a delayed time
 * @param delayMillis the number of milliseconds to delay
 */
export const delayExecution = async <T>(func: () => T, delayMillis = 50): Promise<T> =>
  new Promise<T>((resolve, reject) => {
    setTimeout(() => {
      try {
        resolve(func());
      } catch (e) {
        reject(e);
      }
    }, delayMillis);
  });

/**
 * Calls a function that accepts an arbitrary number of
 * arguments within setTimeout and returns clearTimeout
 *
 * @param func the function to be called
 * @param timeout optional timeout
 * @param funcArgs function arguments
 * @returns
 */
export function delayExecutionReturnClearTimeout(
  func: (...args: any[]) => void,
  timeout = 0,
  ...funcArgs: any[]
): () => void {
  const timer = setTimeout(() => func(...funcArgs), timeout);
  return () => {
    clearTimeout(timer);
  };
}
