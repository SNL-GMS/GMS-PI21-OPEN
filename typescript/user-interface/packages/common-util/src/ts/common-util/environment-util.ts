import type { CommonTypes } from '@gms/common-model';
import { UserProfileTypes } from '@gms/common-model';

import { isWindowDefined } from './window-util';

const windowIsDefined = isWindowDefined();

// !The environment utils currently depends on the common-model for UserMode
// TODO determine if the util library should only use primitive types
// TODO determine if the environment utils should be moved to a different package

/**
 * The NODE_ENV environment variable.
 */
export const { NODE_ENV } = process.env;

/**
 * True if NODE_ENV is set to development.
 */
export const IS_NODE_ENV_DEVELOPMENT = NODE_ENV === 'development';

/**
 * True if NODE_ENV is set to production.
 */
export const IS_NODE_ENV_PRODUCTION = NODE_ENV === 'production';

/**
 * True if NODE_ENV is set to test.
 */
export const IS_NODE_ENV_TEST = NODE_ENV === 'test';

/**
 * The UI Mode ['ian', 'soh' or 'legacy']
 */
export const GMS_UI_MODE = process.env.GMS_UI_MODE || undefined;

/**
 * True if configured for ALL; false otherwise.
 */
export const IS_MODE_LEGACY = GMS_UI_MODE === UserProfileTypes.UserMode.LEGACY;

/**
 * True if configured for IAN; false otherwise.
 */
export const IS_MODE_IAN = GMS_UI_MODE === UserProfileTypes.UserMode.IAN;

/**
 * True if configured for SOH; false otherwise.
 */
export const IS_MODE_SOH = GMS_UI_MODE === UserProfileTypes.UserMode.SOH;

/**
 * The current user mode, which defines which layouts are supported
 */
export const CURRENT_USER_MODE =
  (IS_MODE_LEGACY && UserProfileTypes.UserMode.LEGACY) ||
  (IS_MODE_IAN && UserProfileTypes.UserMode.IAN) ||
  (IS_MODE_SOH && UserProfileTypes.UserMode.SOH) ||
  '';

/** Useful because unit tests can easily override functions. */
export const isIanMode = (): boolean => IS_MODE_IAN;

/** Useful because unit tests can easily override functions. */
export const isSohMode = (): boolean => IS_MODE_SOH;

/**
 * Returns the supported modes based on the current user mode.
 */
export const SUPPORTED_MODES: UserProfileTypes.UserMode[] =
  (CURRENT_USER_MODE === UserProfileTypes.UserMode.SOH && [UserProfileTypes.UserMode.SOH]) ||
  (CURRENT_USER_MODE === UserProfileTypes.UserMode.IAN && [UserProfileTypes.UserMode.IAN]) ||
  Object.keys(UserProfileTypes.UserMode).map(mode => UserProfileTypes.UserMode[mode]);

/**
 * The UI_URL endpoint. This is the URL from which the UI content is served.
 */
export const UI_URL = windowIsDefined
  ? `${window.location.protocol}//${window.location.host}`
  : 'http://localhost:8080';

/**
 * The UI_BASE_PATH endpoint. This is the base path for the URL in a deployment
 */
export const UI_BASE_PATH =
  windowIsDefined && !window.location.host.includes('localhost') ? '/interactive-analysis-ui' : '';

/**
 * The SUBSCRIPTION protocol.
 */
export const UI_SUBSCRIPTION_PROTOCOL =
  windowIsDefined && window.location.protocol === 'https:' ? 'wss' : 'ws';

/**
 * The SUBSCRIPTION URL endpoint.
 */
export const UI_SUBSCRIPTION_URL = windowIsDefined
  ? `${UI_SUBSCRIPTION_PROTOCOL}://${window.location.host}`
  : 'ws://localhost';

/**
 * The GATEWAY_HTTP_PROXY_URI environment variable (or the default value if not set).
 */
export const GATEWAY_HTTP_PROXY_URI = process.env.GATEWAY_HTTP_PROXY_URI || UI_URL;

/**
 * The SUBSCRIPTIONS_PROXY_URI environment variable (or the default value if not set).
 */
export const SUBSCRIPTIONS_PROXY_URI = process.env.SUBSCRIPTIONS_PROXY_URI || UI_SUBSCRIPTION_URL;

/**
 * The API_GATEWAY_URI environment variable (or the default value if not set).
 */
export const API_GATEWAY_URI = GATEWAY_HTTP_PROXY_URI;

/**
 * The API_GATEWAY_URI environment variable for checking a user's login status.
 */
export const API_LOGIN_CHECK_URI = `${GATEWAY_HTTP_PROXY_URI}/interactive-analysis-api-gateway/auth/checkLogIn`;

/**
 * The API_GATEWAY_URI environment variable for accessing the login endpoint.
 */
export const API_LOGIN_URI = `${GATEWAY_HTTP_PROXY_URI}/interactive-analysis-api-gateway/auth/logInUser`;

/**
 * The API_GATEWAY_URI environment variable for accessing the logout endpoint.
 */
export const API_LOGOUT_URI = `${GATEWAY_HTTP_PROXY_URI}/interactive-analysis-api-gateway/auth/logOutUser`;

/**
 * The CESIUM_OFFLINE environment variable.
 */
export const CESIUM_OFFLINE = process.env.CESIUM_OFFLINE
  ? !(
      process.env.CESIUM_OFFLINE === 'null' ||
      process.env.CESIUM_OFFLINE === 'undefined' ||
      process.env.CESIUM_OFFLINE === 'false'
    )
  : false;

/**
 * The `AVAILABLE_SOUND_FILES` environment variable.
 * The available configured sound files for the system.
 */
export const AVAILABLE_SOUND_FILES: string[] =
  process.env.AVAILABLE_SOUND_FILES &&
  process.env.AVAILABLE_SOUND_FILES !== 'undefined' &&
  process.env.AVAILABLE_SOUND_FILES !== 'null'
    ? process.env.AVAILABLE_SOUND_FILES.split(';')
    : [];

export const VERSION_INFO: CommonTypes.VersionInfo = {
  versionNumber: process.env.VERSION_NUMBER ?? 'Unknown version',
  commitSHA: process.env.COMMIT_SHA ?? 'Unknown commit'
};

/**
 * Turns on timing points for UI. Set to 'verbose' to see log timing points.
 * Set to any other string to see warnings level timing only.
 */
export const GMS_PERFORMANCE_MONITORING_ENABLED =
  process.env.GMS_PERFORMANCE_MONITORING_ENABLED?.toLocaleLowerCase() ?? false;
