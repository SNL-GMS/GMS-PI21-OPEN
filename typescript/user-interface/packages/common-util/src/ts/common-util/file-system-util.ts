import parse from 'csv-parse/lib/sync';
import fs from 'fs';

import { jsonPretty } from './json-util';
import { Logger } from './logger';

const logger = Logger.create('GMS_LOG_FILE_SYSTEM_UTIL', process.env.GMS_LOG_FILE_SYSTEM_UTIL);

/**
 * Reads the provided source JSON file into memory
 *
 * @param jsonFilePath The JSON filename from which to read the JSON content
 */
// !FIX: This should be generic and certainly should not always return an array
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function readJsonData<T = any>(jsonFilePath: string): T[] {
  const fileContents = fs.readFileSync(jsonFilePath, 'utf8');
  return JSON.parse(fileContents);
}

/**
 * Write the formatted JSON to file used by other write to file functions
 *
 * @param filePath
 * @param json
 */
export function writeJsonToFile(filePath: string, json: string): void {
  fs.writeFile(filePath, json, err => {
    if (err) {
      logger.warn(err.message);
    }
  });
}
/**
 * Writes provided object to file stringify and pretty
 *
 * @param object object to stringify and written to a file
 * @param fileName filename doNOT include extension
 */
export function writeJsonPretty(object: unknown, fileName: string): void {
  writeJsonToFile(`${fileName}.json`, jsonPretty(object));
}

/**
 * Saves given object to file
 * This will save us time from copying data from the terminal
 *
 * @param object object to save to file
 * @param filePath file path including file name
 */
export function writeObjectToJsonFile(object: unknown, filePath: string): void {
  writeJsonToFile(filePath, JSON.stringify(object, undefined, 2));
}

/** Utility functions for handling CSV files and other file related utils */

/**
 * Reads the provided source CSV file into memory
 *
 * @param csvFilePath
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function readCsvData(csvFilePath: string): any[] {
  const fileContents = fs.readFileSync(csvFilePath, 'utf8');
  return parse(fileContents, { columns: true, delimiter: '\t' });
}

/**
 * Resolves the home value in the config and returns the path
 *
 * @param configString
 */
export function resolveHomeDataPath(configString: string): string[] {
  // Resolve the ${HOME} value or ${SOMETHING} value
  const regex = new RegExp(''.concat('\\$\\{', '([^}]+)', '\\}'), 'g');
  return [configString.replace(regex, (_, v) => process.env[v])];
}
