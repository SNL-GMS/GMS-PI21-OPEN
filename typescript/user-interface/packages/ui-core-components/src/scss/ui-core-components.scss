@import './core-colors';
@import './core-utils';
@import './core-fonts.scss';
@import './components/all';
