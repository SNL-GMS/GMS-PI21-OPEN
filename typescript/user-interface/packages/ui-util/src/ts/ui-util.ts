export * from './ui-util/class-list-util';
export * from './ui-util/color-util';
export * from './ui-util/cookie-util';
export * from './ui-util/coordinate-util';
export * from './ui-util/custom-hooks';
export * from './ui-util/data-attribute-util';
export * from './ui-util/dom-util';
export * from './ui-util/drag-util';
export * from './ui-util/electron-util';
export * from './ui-util/favicon-util';
export * from './ui-util/gl-util';
export * as HotkeyListener from './ui-util/hot-key-util';
export * from './ui-util/logger';
export * from './ui-util/os-util';
export * from './ui-util/search-util';
export * from './ui-util/selection-util';
export * from './ui-util/time-util';
