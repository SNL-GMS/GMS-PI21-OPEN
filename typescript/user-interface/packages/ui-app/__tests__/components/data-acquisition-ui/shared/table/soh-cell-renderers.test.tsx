/* eslint-disable @typescript-eslint/no-unused-vars */
import { uuid, ValueType } from '@gms/common-util';
import type { PercentBarProps } from '@gms/ui-core-components';
import { PercentBar } from '@gms/ui-core-components';
import { UILogger } from '@gms/ui-util';
import { render } from '@testing-library/react';
import React from 'react';

import type {
  NameCellProps,
  SohCellRendererProps
} from '../../../../../src/ts/components/data-acquisition-ui/shared/table/soh-cell-renderers';
import {
  formatSohValue,
  NameCell,
  SohRollupCell
} from '../../../../../src/ts/components/data-acquisition-ui/shared/table/soh-cell-renderers';
import {
  CellStatus,
  DataReceivedStatus
} from '../../../../../src/ts/components/data-acquisition-ui/shared/table/utils';

const logger = UILogger.create(
  'GMS_SOH_CELL_RENDERERS_TEST',
  process.env.GMS_SOH_CELL_RENDERERS_TEST
);

// eslint-disable-next-line @typescript-eslint/no-var-requires, @typescript-eslint/no-require-imports
const Enzyme = require('enzyme');

// set up window alert and open so we don't see errors
(window as any).alert = jest.fn();
(window as any).open = jest.fn();

uuid.asString = jest.fn().mockReturnValue('1e872474-b19f-4325-9350-e217a6feddc0');

/**
 * Tests the ability to check if the peak trough is in warning
 */
describe('Station cell renderers', () => {
  // start formatSohDetailValue function tests
  test('formatSohDetailValue function can detect non-numbers', () => {
    expect(formatSohValue('NaN' as any, ValueType.PERCENTAGE, true)).toEqual('Unknown');
  });

  test('formatSohDetailValue function correctly formats input number', () => {
    const inputNumber = 999;
    expect(formatSohValue(inputNumber, ValueType.PERCENTAGE, true)).toEqual('999.00');
  });
  // end formatSohDetailValue function tests

  // start Percentage bar pure component tests
  const percentBarProps: PercentBarProps = {
    percentage: 10
  };
  // eslint-disable-next-line react/jsx-props-no-spreading
  const percentBar = Enzyme.mount(<PercentBar {...percentBarProps} />);
  // TODO redo test in RTL
  test('PercentBar should be defined', () => {
    expect(percentBar).toBeDefined();
  });
  test('PercentBar should not have state', () => {
    expect(percentBar.state()).toEqual(null);
  });
  test('PercentBar should have props of type PercentBarProps', () => {
    expect(percentBar.props()).toEqual({ percentage: 10 });
  });
  test('PercentBar match snapshot', () => {
    expect(percentBar).toMatchSnapshot();
  });
  // end Percentage bar tests

  // start StationCellRenderer function component tests
  const stationCellRendererProps: SohCellRendererProps = {
    className: 'string',
    stationId: 'string',
    value: 'string',
    dataReceivedStatus: DataReceivedStatus.RECEIVED,
    cellStatus: CellStatus.GOOD
  };
  const nameCellProps: NameCellProps = {
    value: 'name',
    valueFormatted: 'NAME',
    data: undefined,
    node: undefined,
    name: 'test',
    colDef: undefined,
    api: undefined,
    columnApi: undefined,
    context: undefined,
    getValue: jest.fn(),
    setValue: jest.fn(),
    formatValue: jest.fn(),
    column: undefined,
    rowIndex: 1,
    refreshCell: undefined,
    eGridCell: undefined,
    eParentOfValue: undefined,
    cellStatus: CellStatus.GOOD,
    dataReceivedStatus: DataReceivedStatus.RECEIVED,
    registerRowDragger(
      rowDraggerElement: HTMLElement,
      dragStartPixels?: number,
      value?: string
    ): void {
      logger.info('registerRowDragger function not implemented');
    }
  };
  // eslint-disable-next-line react/jsx-props-no-spreading
  const stationCellRenderer = Enzyme.mount(<SohRollupCell {...stationCellRendererProps} />);
  // eslint-disable-next-line react/jsx-props-no-spreading
  const nameCell = Enzyme.mount(<NameCell {...nameCellProps} />);
  // TODO redo test in RTL
  it('stationCellRenderer should be defined', () => {
    expect(stationCellRenderer).toBeDefined();
  });
  it('NameCell should be defined', () => {
    expect(NameCell).toBeDefined();
  });
  it('stationCellRenderer should have props of type stationCellRendererProps', () => {
    expect(stationCellRenderer.props()).toMatchSnapshot();
  });

  it('nameCell should have props of type nameCellProps', () => {
    expect(nameCell.props()).toMatchSnapshot();
  });

  it('stationCellRenderer should match snapshot', () => {
    // eslint-disable-next-line react/jsx-props-no-spreading
    const { container } = render(<SohRollupCell {...stationCellRendererProps} />);
    expect(container).toMatchSnapshot();
  });
  // end StationCellRenderer function component tests
});
