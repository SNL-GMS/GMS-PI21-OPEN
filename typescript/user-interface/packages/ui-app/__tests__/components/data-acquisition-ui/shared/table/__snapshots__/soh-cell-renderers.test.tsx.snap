// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Station cell renderers PercentBar match snapshot 1`] = `
<PercentBar
  percentage={10}
>
  <div
    className="percent-bar"
    key="1"
    style={
      Object {
        "transform": "translateX(-90%)",
        "width": "100%",
      }
    }
  />
</PercentBar>
`;

exports[`Station cell renderers nameCell should have props of type nameCellProps 1`] = `
Object {
  "api": undefined,
  "cellStatus": "good",
  "colDef": undefined,
  "column": undefined,
  "columnApi": undefined,
  "context": undefined,
  "data": undefined,
  "dataReceivedStatus": "received",
  "eGridCell": undefined,
  "eParentOfValue": undefined,
  "formatValue": [MockFunction],
  "getValue": [MockFunction],
  "name": "test",
  "node": undefined,
  "refreshCell": undefined,
  "registerRowDragger": [Function],
  "rowIndex": 1,
  "setValue": [MockFunction],
  "value": "name",
  "valueFormatted": "NAME",
}
`;

exports[`Station cell renderers stationCellRenderer should have props of type stationCellRendererProps 1`] = `
Object {
  "cellStatus": "good",
  "className": "string",
  "dataReceivedStatus": "received",
  "stationId": "string",
  "value": "string",
}
`;

exports[`Station cell renderers stationCellRenderer should match snapshot 1`] = `
<div>
  <div
    class="table-cell soh-cell string"
    data-cell-status="good"
    data-cy="table-cell"
    data-received-status="received"
    data-station-id="string"
    style="height: 36px; overflow: hidden;"
    title="string"
  >
    <div
      class=" table-cell__value "
      data-cy="table-cell__value"
    >
      <span>
        string
      </span>
    </div>
  </div>
</div>
`;
